### PR TITLE
Formalize zero-width emote implementation

### LINKED ISSUE #2994: Animated emotes renders on top of emote modifiers.
**Describe your issue**
Animated emotes renders on top of emote modifiers.
Expected behavior: Have the animated emotes below emote modifiers, just like normal emotes.

**Screenshots**
Chatterino:
Chatterino Screenshot
Web Chat (FFZ extension):
Chatterino Screenshot

**OS and Chatterino Version**
Windows 10, Chatterino Nightly 2.3.2

### LINKED ISSUE #3615: Tooltips for zero-width emotes are inaccessable
### Checklist

- [X] I'm reporting a problem with Chatterino
- [X] I've verified that I'm running **the most** recent nightly build or stable release
- [X] I've looked for my problem on the [wiki](https://wiki.chatterino.com/Help/)
- [X] I've searched the [issues and pull requests](https://github.com/Chatterino/chatterino2/issues?q=) for similar looking reports

### Describe your issue

When a zero-width emote is posted along side a standard emote, only a tooltip for the standard emote is shown no matter where the mouse is hovered. 

### Screenshots

N/A

### OS and Chatterino Version

Chatterino 7.3.4 (commit 530dcc3c), applies to c2

### PR DESCRIPTION
**Pull request by dnsge**
Pull request checklist:

- [x] `CHANGELOG.md` was updated, if applicable

# Description

This PR attempts to fix the notion of a zero-width emote by introducing layered emotes: when one (or more) zero-width emotes follow a regular emote, the emotes are all gathered into a single `LayeredEmoteElement` which handles layered rendering through the `LayeredImageLayoutElement` and provides useful tooltips.

The `TooltipWidget` now has two "styles": a vertical layout, and a grid layout. If there are more than 2 emotes being shown at once in the tooltip, it will use the grid layout. Otherwise, the vertical layout will be used.

<details>
<summary>Images of changes</summary>
(margins and spacing are slightly out of date as of 62edb0c)

**Single zero-width emote:**
Screen Shot 2023-01-18 at 8 57 04 PM

**Multiple zero-width emotes:**
Screen Shot 2023-01-18 at 8 57 18 PM

**Excessive number of zero-width emotes:**
Screen Shot 2023-01-18 at 8 57 33 PM

**Context menu with zero-width emotes:**
Screen Shot 2023-01-18 at 9 00 48 PM

</details>

Also adds a "Enable zero-width emotes" setting (default: true). If off, zero-width emotes are shown as regular emotes.

Fixes #3615
Fixes #2994

### CODE REVIEW

**Comment by Felanbird** [comment]
Functionality looks good from brief testing, will use this build for the week and see if I run into any issues 👍 

Also as noted by my linked-issue, this PR resolves the issues with animated BTTV emotes rendering on-top of zero-width emotes

image

**Comment by Felanbird** [comment]
First thing I've noticed is the general tool-tip size is larger, much more noticeable by the space in-between the emote info and the bottom

Before:
image
image


Now:
image
image

**Comment by dnsge** [comment]
> general tool-tip size is larger

Should be fixed now in 62edb0c

File: src/widgets/TooltipEntryWidget.hpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    bool refreshPixmap();
+
+    bool animated() const;
+    ImagePtr getImage() const;
+
+private:
+    QLabel *displayImage_;
+    QLabel *displayText_;
+
+    bool attemptRefresh{false};
```
Comment (kornes) [inline src/widgets/TooltipEntryWidget.hpp:36]: unused leftover

File: src/widgets/TooltipWidget.cpp [hunk 2]
```diff
@@ -49,123 +41,260 @@ TooltipWidget::TooltipWidget(BaseWidget *parent)
...
-        if (this->image_ && this->attemptRefresh)
+        for (int i = 0; i < this->visibleEntries_; ++i)
         {
-            if (this->refreshPixmap())
+            auto entry = this->entryAt(i);
+            if (entry->getImage())
             {
-                this->attemptRefresh = false;
-                this->adjustSize();
+                if (entry->refreshPixmap())
```
Comment (kornes) [inline src/widgets/TooltipWidget.cpp:60]: miscUpdate ticks every 100ms, so preferably refreshPixmap should be called only as long it returns false for given entry

**Review by kornes — COMMENTED** [review]
very nice feature, gonna see how it runs for few days 👍

**Comment by Felanbird** [comment]
> > general tool-tip size is larger
> 
> Should be fixed now in [62edb0c](https://github.com/Chatterino/chatterino2/commit/62edb0c00a13897b1e8ddaae786526ce639b8770)

Issue is resolved, there's a slight miscentering when there's exactly 2 zero-width emotes, but I'd take that over the huge tool-tips any day 👍

File: src/widgets/TooltipWidget.cpp (hunk 2 above)
Comment (dnsge) [reply src/widgets/TooltipWidget.cpp]: Good point, updated in [a275d67](https://github.com/Chatterino/chatterino2/pull/4314/commits/a275d67ba6a8d9c6ba6ffbb1817712bb5d110942)

File: src/widgets/settingspages/GeneralPage.cpp [hunk 3]
```diff
@@ -374,6 +374,7 @@ void GeneralPage::initLayout(GeneralPageView &layout)
     layout.addCheckbox("Animate", s.animateEmotes);
     layout.addCheckbox("Animate only when Chatterino is focused",
                        s.animationsWhenFocused);
+    layout.addCheckbox("Enable zero-width emotes", s.enableZeroWidthEmotes);
```
Comment (Felanbird) [inline src/widgets/settingspages/GeneralPage.cpp:377]:
```suggestion
    layout.addCheckbox(
        "Enable zero-width emotes", s.enableZeroWidthEmotes, false,
        "When disabled, emotes that overlap other emotes such as BTTV's cvMask "
        "and 7TV's RainTime, will appear as normal emotes.");
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -10,6 +10,7 @@
 - Minor: Added link to streamlink docs for easier user setup. (#4217)
 - Minor: Added setting to turn off rendering of reply context. (#4224)
 - Minor: Added setting to select which channels to log. (#4302)
+- Minor: Improved zero-width emote tooltips and rendering. (#4314)
```
Comment (Felanbird) [inline CHANGELOG.md:13]: maybe better as bugfix changelogs?

`- Bugfix: Fixed an issue where animated BTTV emotes would render on top of zero-width emotes (#4314)`
`- Bugfix: Fixed an issue where it was difficult to hover a zero-width emote. (#4314)`

Comment (dnsge) [reply CHANGELOG.md]: Agreed

**Review by Felanbird — APPROVED** [review]
I believe this works as expected 👍 - I actually feel like I've seen less zero-width emotes since I started using this PR, but it's much more likely less are being used than something being broken.

File: src/widgets/helper/ChannelView.cpp [hunk 5]
```diff
@@ -2134,6 +2196,36 @@ void ChannelView::addImageContextMenuItems(
             addEmoteContextMenuItems(*emoteElement->getEmote(), creatorFlags,
                                      menu);
         }
+        else if (auto layeredElement =
+                     dynamic_cast<const LayeredEmoteElement *>(&creator))
+        {
+            // Functor for std::copy_if that keeps track of seen elements
```
Comment (pajlada) [inline src/widgets/helper/ChannelView.cpp:2202]: I'd prefer if "get unique emotes" is added as a helper function to LayeredEmoteElement instead

File: src/widgets/helper/ChannelView.cpp [hunk 6]
```diff
@@ -1658,37 +1659,94 @@ void ChannelView::mouseMoveEvent(QMouseEvent *event)
...
+                        element->getTooltip(),
+                    });
+                }
+                else if (layeredEmoteElement)
+                {
+                    auto &layeredEmotes = layeredEmoteElement->getEmotes();
+                    // Should never be empty but ensure it
+                    if (!layeredEmotes.empty())
+                    {
+                        std::vector<TooltipEntry> records;
```
Comment (pajlada) [inline src/widgets/helper/ChannelView.cpp:1694]: We call many of the variables that have to do with TooltipEntries `record` or `records`, I would've called them `entry` and `entries`

File: src/widgets/helper/ChannelView.cpp [hunk 7]
```diff
@@ -1658,37 +1659,94 @@ void ChannelView::mouseMoveEvent(QMouseEvent *event)
...
+                            upperLimit = TOOLTIP_EMOTE_ENTRIES_LIMIT - 1;
+                            truncating = true;
+                        }
+
+                        for (size_t i = 0; i < upperLimit; ++i)
+                        {
+                            auto &emote = layeredEmotes[i];
+                            if (i == 0)
+                            {
+                                // Large image and full description
```
Comment (pajlada) [inline src/widgets/helper/ChannelView.cpp:1716]: nit: // The first entry gets...

File: src/widgets/helper/ChannelView.cpp [hunk 8]
```diff
@@ -1658,37 +1659,94 @@ void ChannelView::mouseMoveEvent(QMouseEvent *event)
...
+                                records.push_back({emote->images.getImage(3.0),
+                                                   emoteTooltips[i]});
+                            }
+                            else
+                            {
+                                // Smaller image and just emote name
```
Comment (pajlada) [inline src/widgets/helper/ChannelView.cpp:1722]: nit: // All remaining entries get...

File: src/messages/MessageElement.hpp [hunk 9]
```diff
@@ -141,9 +141,7 @@ enum class MessageElementFlag : int64_t {
     LowercaseLink = (1LL << 29),
     OriginalLink = (1LL << 30),
 
-    // ZeroWidthEmotes are emotes that are supposed to overlay over any pre-existing emotes
-    // e.g. BTTV's SoSnowy during christmas season or 7TV's RainTime
-    ZeroWidthEmote = (1LL << 31),
+    // Unused: (1LL << 31)
```
Comment (pajlada) [inline src/messages/MessageElement.hpp:144]: I am fine with this :+1:

File: src/messages/MessageElement.hpp [hunk 10]
```diff
@@ -321,6 +319,38 @@ class EmoteElement : public MessageElement
     EmotePtr emote_;
 };
 
+class LayeredEmoteElement : public MessageElement
```
Comment (pajlada) [inline src/messages/MessageElement.hpp:322]: Could you add a class comment for this? It should include what purpose it has/what it tries to solve, not necessarily any technical details for how it solves it.

File: src/messages/MessageElement.hpp [hunk 11]
```diff
@@ -321,6 +319,38 @@ class EmoteElement : public MessageElement
...
+    LayeredEmoteElement(
+        std::vector<EmotePtr> &&emotes, MessageElementFlags flags,
+        const MessageColor &textElementColor = MessageColor::Text);
+
+    void addEmoteLayer(const EmotePtr &emote);
+
+    void addToContainer(MessageLayoutContainer &container,
+                        MessageElementFlags flags) override;
+
+    QString getCleanCopyString() const;
```
Comment (pajlada) [inline src/messages/MessageElement.hpp:334]: Comment this function

**Review by pajlada — CHANGES_REQUESTED** [review]
I've taken a bit of a glance at the code, overall looks fine - mostly nitpicks in my review.
Thank you for looking into this, this feels like an overall better & more consistent approach!

File: src/widgets/helper/ChannelView.cpp (hunk 6 above)
Comment (dnsge) [reply src/widgets/helper/ChannelView.cpp:1694]: Good catch -- while I was developing this PR I changed the name from `TooltipRecord` to `TooltipEntry` but forgot to update variable names

File: src/widgets/TooltipEntryWidget.hpp [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+class TooltipEntryWidget : public QWidget
+{
+    Q_OBJECT
+
+public:
+    TooltipEntryWidget(QWidget *parent = nullptr);
+    TooltipEntryWidget(ImagePtr image, const QString &text,
+                       QWidget *parent = nullptr);
+    TooltipEntryWidget(ImagePtr image, const QString &text, int customWidth,
```
[the remark this answers is not in the document]
Comment (dnsge) [reply src/widgets/TooltipEntryWidget.hpp]: Not sure if I understand. I did find that the second constructor is actually unused and removed it in 51ff1c3e. However, the `TooltipEntryWidget` exists outside of the grid/vertical layout, e.g. the `TooltipEntryWidget` is inserted into the appropriate layout determined by `TooltipWidget`.

I suppose that an argument could be made that `TooltipEntryWidget` should be moved inside `TooltipWidget` as nothing else should need to access `TooltipEntryWidget`? I don't really mind either way.

**Review by pajlada — APPROVED** [review]
@dnsge If you could take a last look at your own code & the changes I've made before I merge this in that would be great - seems to work as expected for me with tooltips & everything on both X11 & Sway

**Pull request merged**